### PR TITLE
Account for forwards which whose corresponding backwards are not invoked

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98341
* #98254
* #97440
* __->__ #98112
* #98101
* #98097

Previously, when we would run a forward graph whose backward we never invoked it would prevent us from switching from warmup to recording. Now, refine the heuristic to allow incrementing the generation as soon as we invoke a backward graph. This still handles the 
```
mod1 = torch.compile(...)

mod2 = torch.compile(...) 

mod2(mod1(x)).sum().backward()
```
case while accounting for graphs which we may not run backward of. 

It also now handles the case where we skip cudagraphify the backward of a forward.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
Test?

**Comment by ezyang** [comment]
i guess you removed review request so imma wait

**Comment by eellison** [comment]
@ezyang, yea, adding test

**Review by jansel — APPROVED** [review]

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -215,6 +225,13 @@ def compile_fx_inner(
             and not complex_memory_overlap_inputs
             and (len(graph.device_idxs) == 1 or not config.triton.cudagraph_trees)
         ):
+            if (
+                boxed_forward_device_index is not None
+                and not is_inference
+                and not is_backward
+            ):
+                boxed_forward_device_index.set(next(iter(graph.device_idxs)))
```
Comment (ezyang) [inline torch/_inductor/compile_fx.py:233]: Why is it correct to only take the first device index?

**Pull request closed without merging**